---
title: "ChIP-seq Results Report"
author: "Jianhong Ou"
date: "`r date()`"
output: 
  html_document:
    theme: simplex
    toc: true
    toc_float: true
    toc_depth: 4
params:
  design: ""
  genome: ""
  summary: ""
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DT)
library(yaml)
```

# ![QiuBio Pipeline/MintChIP](images/mintchiplogo.png)

This document describes the output produced by the MintChIP analysis pipeline. Most of the plots are taken from the [MultiQC report](multiqc/multiqc_report.html), which summaries results at the end of the pipeline.

## Experiment design

```{r, results = 'asis', echo=FALSE}
data <- yaml::yaml.load_file(params$summary)
cat(data$data, sep = "\n")
```

### Design table

The full design table header can be 'group', 'replicate', 'fastq_1', 'fastq_2', 'antibody', 'control', 'md5_1', 'md5_2', 'peaktype', and 'track_...' (prefix with track_).
If 'md5_1' and/or 'md5_2' are set, the pipeline will double check the fastq_1 and fastq_2 files.
If 'peaktype' is set, the pipeline will use the peaktype to call peaks.

```{r, echo=FALSE}
data <- read.csv(params$design)
datatable(data)
```


## Pipeline overview

The pipeline is built based on [nf-core/chipseq](https://github.com/nf-core/chipseq) using [Nextflow](https://www.nextflow.io/). See [`main README.md`](https://github.com/jianhong/MintChiP) for a condensed overview of the steps in the pipeline, and the bioinformatics tools used at each step.

See [Illumina website](https://emea.illumina.com/techniques/sequencing/dna-sequencing/chip-seq.html) for more information regarding the ChIP-seq protocol, and for an extensive list of publications.

This index file will be created in the output directory after the pipeline has finished. 
See [`results description`](pipeline_info/results_description.html) for full description of the output files for the pipeline.

## QC metrics

### library preparation and sequence quality

[Picard MergeSamFiles and MarkDuplicates](https://broadinstitute.github.io/picard/command-line-overview.html) are used in combination to merge the alignments, and for the marking of duplicates, respectively. If you only have one library for any given replicate then the merging step isn't carried out because the library-level and merged library-level BAM files will be exactly the same.

![MultiQC - Picard deduplication stats plot](`r ifelse(file.exists("multiqc_plots/png/mqc_picard_deduplication_1.png"), "multiqc_plots/png/mqc_picard_deduplication_1.png", "images/mqc_picard_deduplication_plot.png")`)

Read duplicate marking is carried out using the Picard MarkDuplicates command. Duplicate reads are generally removed from the aligned reads to mitigate for fragments in the library that may have been sequenced more than once due to PCR biases. A selection of alignment-based QC metrics generated by Picard CollectMultipleMetrics and MarkDuplicates will be included in the MultiQC report.

![MultiQC - Picard insert size plot](`r ifelse(file.exists("multiqc_plots/png/mqc_picard_insert_size_Counts.png"), "multiqc_plots/png/mqc_picard_insert_size_Counts.png", "images/mqc_picard_insert_size_plot.png")`)

The [Preseq](http://smithlabresearch.org/software/preseq/) package is aimed at predicting and estimating the complexity of a genomic sequencing library, equivalent to predicting and estimating the number of redundant reads from a given sequencing depth and how many will be expected from additional sequencing using an initial sequencing experiment. The estimates can then be used to examine the utility of further sequencing, optimize the sequencing depth, or to screen multiple libraries to avoid low complexity samples. The dashed line shows a perfectly complex library where total reads = unique reads. Note that these are predictive numbers only, not absolute. The MultiQC plot can sometimes give extreme sequencing depth on the X axis - click and drag from the left side of the plot to zoom in on more realistic numbers.

![MultiQC - Preseq library complexity plot](`r ifelse(file.exists("multiqc_plots/png/mqc_preseq_plot_1.png"), "multiqc_plots/png/mqc_preseq_plot_1.png", "images/mqc_preseq_plot.png")`)


[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/) gives general quality metrics about your reads. It provides information about the quality score distribution across your reads, the per base sequence content (%A/C/G/T). You get information about adapter contamination and other over-represented sequences.

![MultiQC - FastQC per base sequence quality plot](`r ifelse(file.exists("multiqc_plots/png/mqc_fastqc_per_base_sequence_quality_plot_1.png"), "multiqc_plots/png/mqc_fastqc_per_base_sequence_quality_plot_1.png", "https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc.png")`)

For further reading and documentation see the [FastQC help](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

> **NB:** The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality. To see how your reads look after trimming, look at the FastQC reports in the `trim_galore` directory.


### ChIP-seq QC

<details markdown="1">
    <summary>Output files</summary>

* `bwa/mergedLibrary/phantompeakqualtools/`
    * `*.spp.out`, `*.spp.pdf`: phantompeakqualtools output files.
    * `*_mqc.tsv`: MultiQC custom content files.
* `bwa/mergedLibrary/deepTools/plotFingerprint/`
    * `*.plotFingerprint.pdf`, `*.plotFingerprint.qcmetrics.txt`, `*.plotFingerprint.raw.txt`: plotFingerprint output files.
* `bwa/mergedLibrary/deepTools/plotProfile/`
    * `*.computeMatrix.mat.gz`, `*.computeMatrix.vals.mat.tab`, `*.plotProfile.pdf`, `*.plotProfile.tab`, `*.plotHeatmap.pdf`, `*.plotHeatmap.mat.tab`: plotProfile output files.

</details>

```{r, echo=FALSE}
cvgs <- dir("multiqc_plots/svg", "mplplot", full.names = TRUE)
cvgs <- cvgs[!grepl("_pc.svg", cvgs)]
files <- c(spp="Strand-shift",
           NSC="NSC coefficient",
           RSC="RSC coefficient",
           FRiP="FRiP score",
           Peak="Peak count",
           Feature="Feature")
content <- lapply(cvgs, readLines, warn = FALSE)
filename <- sapply(files, function(.ele) {
  con <- sapply(content, function(.e) any(grepl(.ele, .e)))
  if(any(con)){
    cvgs[con][1]
  }else{
    "N/A"
  }
})
filename <- gsub("svg", "png", filename)
```

## I need a table here to replace the figures.

Normalized strand coefficient (NSC) is the normalized ratio between the fragment-length cross-correlation peak and the background cross-correlation. NSC values range from a minimum of 1 to larger positive numbers. 1.1 is the critical threshold. Dataset with NSC values much less than 1.1 (< 1.05) tend to have low signal to noise or few peaks (this could be biological e.g. a factor that truly binds only a few sites in a particular tissue type OR it could be due to poor quality). ENCODE cut-off: **NSC > 1.05**.

![MultiQC - spp NSC plot](`r ifelse(file.exists(filename["NSC"]), filename["NSC"], "images/mqc_spp_nsc_plot.png")`)

Relative strand correlation (RSC) is the ratio between the fragment-length peak and the read-length peak. RSC values range from 0 to larger positive values. 1 is the critical threshold. RSC values significantly lower than 1 (< 0.8) tend to have low signal to noise. The low scores can be due to failed and poor quality ChIP, low read sequence quality and hence lots of mis-mappings, shallow sequencing depth (significantly below saturation) or a combination of these. Like the NSC, dataset with few binding sites (< 200), which is biologically justifiable, also show low RSC scores. ENCODE cut-off: **RSC > 0.8**.

![MultiQC - spp RSC plot](`r ifelse(file.exists(filename["RSC"]), filename["RSC"], "images/mqc_spp_rsc_plot.png")`)

[deepTools](https://deeptools.readthedocs.io/en/develop/content/list_of_tools.html) plotFingerprint is a useful QC for ChIP-seq data in order to see the relative enrichment of the IP samples with respect to the controls on a genome-wide basis. The results, however, are expected to look different for example when comparing narrow marks such as transcription factors and broader marks such as histone modifications (see [plotFingerprint docs](https://deeptools.readthedocs.io/en/develop/content/tools/plotFingerprint.html)).

![MultiQC - deepTools plotFingerprint plot](`r ifelse(file.exists("multiqc_plots/png/mqc_deeptools_fingerprint_plot_1.png"), "multiqc_plots/png/mqc_deeptools_fingerprint_plot_1.png", "images/mqc_deeptools_plotFingerprint_plot.png")`)

Various QC plots per sample including number of peaks, fold-change distribution, [FRiP score](https://genome.cshlp.org/content/22/9/1813.full.pdf+html) and peak-to-gene feature annotation are also generated by the pipeline. Where possible these have been integrated into the MultiQC report.

![MultiQC - MACS2 peaks FRiP score plot](`r ifelse(file.exists(filename["FRiP"]), filename["FRiP"], "images/mqc_frip_score_plot.png")`)


## Signals

### Mapping summary

The mapping rate and stats are listed as following table:


### Alignment

<details markdown="1">
    <summary>Output files</summary>

* `bwa/mergedLibrary/`
    * `*.bam`: Merged library-level, coordinate sorted `*.bam` files after the marking of duplicates, and filtering based on various criteria. The file suffix for the final filtered files will be `*.mLb.clN.*`. If you specify the `--save_align_intermeds` parameter then two additional sets of files will be present. These represent the unfiltered alignments with duplicates marked (`*.mLb.mkD.*`), and in the case of paired-end datasets the filtered alignments before the removal of orphan read pairs (`*.mLb.flT.*`).
* `bwa/mergedLibrary/samtools_stats/`
    * SAMtools `*.flagstat`, `*.idxstats` and `*.stats` files generated from the alignment files.
* `bwa/mergedLibrary/picard_metrics/`
    * `*_metrics`: Alignment QC files from picard CollectMultipleMetrics.
    * `*.metrics.txt`: Metrics file from MarkDuplicates.
* `bwa/mergedLibrary/picard_metrics/pdf/`
    * `*.pdf`: Alignment QC plot files from picard CollectMultipleMetrics.
* `bwa/mergedLibrary/preseq/`
    * `*.ccurve.txt`: Preseq expected future yield file.

> **NB:** File names in the resulting directory (i.e. `bwa/mergedLibrary/`) will have the '`.mLb.`' suffix.

</details>

Adapter-trimmed reads are mapped to the reference assembly using [BWA](http://bio-bwa.sourceforge.net/bwa.shtml). 
BWA is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome.
BWA-MEM is recommended for high-quality queries and over 70bp reads. 
If the reads is smaller than 50bp, this pipeline may be **NOT** accurate.


### Normalised bigWig files

<details markdown="1">
    <summary>Output files</summary>

* `bwa/mergedLibrary/bigwig/`
    * `*.bigWig`: Normalised bigWig files scaled to 1 million mapped reads.
* `bwa/mergedLibrary/bigwig/deepTools`
    * `*.bw`: Normalised bigWig files scaled to 1 million mapped reads or per genome size.
    
</details>

The [bigWig](https://genome.ucsc.edu/goldenpath/help/bigWig.html) format is in an indexed binary format useful for displaying dense, continuous data in Genome Browsers such as the [UCSC](https://genome.ucsc.edu/cgi-bin/hgTracks) and [IGV](http://software.broadinstitute.org/software/igv/). This mitigates the need to load the much larger BAM files for data visualisation purposes which will be slower and result in memory issues. The coverage values represented in the bigWig file can also be normalized in order to be able to compare the coverage across multiple samples - this is not possible with BAM files. The bigWig format is also supported by various bioinformatics software for downstream processing such as meta-profile plotting. The normalized bigWig files are also available in [deepTools](https://deeptools.readthedocs.io/en/develop/content/list_of_tools.html) subfolder.

### IGV session file

<details markdown="1">
    <summary>Output files</summary>

* `igv/<PEAK_TYPE>/`
    * `igv_session.xml`: Session file that can be directly loaded into IGV.
    * `igv_files.txt`: File containing a listing of the files used to create the IGV session.

</details>

An [IGV](https://software.broadinstitute.org/software/igv/UserGuide) session file will be created at the end of the pipeline containing the normalized bigWig tracks, per-sample peaks, consensus peaks and differential sites. This avoids having to load all of the data individually into IGV for visualization.

The genome fasta file required for the IGV session will be the same as the one that was provided to the pipeline. This will be copied into `genome/` to overcome any loading issues. If you prefer to use another path or an in-built genome provided by IGV just change the `genome` entry in the second-line of the session file.

The file paths in the IGV session file will only work if the results are kept in the same place on your storage. If the results are moved or for example, if you prefer to load the data over the web then just replace the file paths with others that are more appropriate.

Once installed, open IGV, go to `File > Open Session` and select the `igv_session.xml` file for loading.

![IGV screenshot](images/igv_screenshot.png)

> **NB:** If you are not using an in-built genome provided by IGV you will need to load the annotation yourself e.g. in .gtf and/or .bed format.

### UCSC genome track hub files

The [UCSC genome track hub](https://genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html) files are generated by [trackhub](https://daler.github.io/trackhub/).
The "composite track", such as group and colors, can be set in the design table by setting the prefix of column name with "track_".

The files comprising a track hub need to be uploaded to a server, 
and a UCSC Genome Browser is pointed to the served URL for viewing. 


## Call peaks

<details markdown="1">
    <summary>Output files</summary>

* `bwa/mergedLibrary/macs/<PEAK_TYPE>/`
    * `*.xls`, `*.broadPeak` or `*.narrowPeak`, `*.gappedPeak`, `*summits.bed`: MACS2 output files - the files generated will depend on whether MACS2 has been run in *narrowPeak* or *broadPeak* mode.
    * `*.annotatePeaks.txt`: HOMER peak-to-gene annotation file.
* `bwa/mergedLibrary/macs/<PEAK_TYPE>/qc/`
    * `macs_peak.plots.pdf`: QC plots for MACS2 peaks.
    * `macs_annotatePeaks.plots.pdf`: QC plots for peak-to-gene feature annotation.
    * `*.FRiP_mqc.tsv`, `*.count_mqc.tsv`, `macs_annotatePeaks.summary_mqc.tsv`: MultiQC custom-content files for FRiP score, peak count and peak-to-gene ratios.

> **NB:** `<PEAK_TYPE>` in the directory structure above corresponds to the type of peak that you have specified to call with MACS2 i.e. `broadPeak` or `narrowPeak`.

</details>

[MACS2](https://github.com/taoliu/MACS) is one of the most popular peak-calling algorithms for ChIP-seq data.  See [MACS2 outputs](https://github.com/taoliu/MACS#output-files) for a description of the output files generated by MACS2.
> **NB:** If the sample name is not listed in the MACS2 total peak count plot or Homer annotatePeaks peak-to-gene feature ratio plot, the peak calling is failed in that sample.

![MultiQC - MACS2 total peak count plot](`r ifelse(file.exists("multiqc_plots/png/mqc_macs2_peak_count_plot.png"), "multiqc_plots/png/mqc_macs2_peak_count_plot.png", "images/mqc_macs2_peak_count_plot.png")`)


## Annotation

The results from [deepTools](https://deeptools.readthedocs.io/en/develop/content/list_of_tools.html) plotProfile gives you a quick visualization for the genome-wide enrichment of your samples at the TSS, and across the gene body. During the downstream analysis, you may want to refine the features/genes used to generate these plots in order to see a more specific condition-related effect.

![MultiQC - deepTools plotProfile plot](`r ifelse(file.exists("multiqc_plots/png/mqc_read_distribution_profile_1.png"), "multiqc_plots/png/mqc_read_distribution_profile_1.png", "images/mqc_deeptools_plotProfile_plot.png")`)


[HOMER annotatePeaks.pl](http://homer.ucsd.edu/homer/ngs/annotation.html) is used to annotate the peaks relative to known genomic features. Please note that some of the output columns will be blank because the annotation is not provided using HOMER's in-built database format. However, the more important fields required for downstream analysis will be populated i.e. *Annotation*, *Distance to TSS* and *Nearest Promoter ID*.
For differential analysis, the peaks will be also annotated with [ChIPpeakAnno package](	https://bioconductor.org/packages/ChIPpeakAnno/), see details in section [differential by DiffBind](#methods-2-differential-by-diffbind).

![MultiQC - HOMER annotatePeaks peak-to-gene feature ratio plot](`r ifelse(file.exists(filename["Peak"]),filename["Peak"], "images/mqc_annotatePeaks_feature_percentage_plot.png")`)

## Differential analysis

Differential analysis were done by three steps:

1. Create pool of peaks from all samples

2. Read counting

3. Differential binding analysis by raw counts

This pipeline will use two methods to do differential analysis. 
The major difference between the two methods are in the peak pooling and read counting step.
The differential analysis are both use a standardized DESeq2 analysis script to get 
an idea of the reproducibility within the experiment, and to assess the overall differential binding. 

The two methods are:

1. Merge peaks by BEDtools::mergeBed, count reads by featureCounts, and 
do differential binding analysis by DESeq2.

2. Merge peaks, count reads and do differential binding analysis by DiffBind.

### Methods 1: differential by DESeq2 for consensus

<details markdown="1">
    <summary>Output files</summary>

* `bwa/mergedLibrary/macs/<PEAK_TYPE>/consensus/<ANTIBODY>/deseq2/`
    * `*.results.txt`: Spreadsheet containing differential binding results across all consensus peaks and all comparisons.
    * `*.plots.pdf`: File containing PCA and hierarchical clustering plots.
    * `*.log`: Log file with information for number of differentially bound intervals at different FDR and fold-change thresholds for each comparison.
    * `*.dds.rld.RData`: File containing R `dds` and `rld` objects generated by DESeq2.
    * `R_sessionInfo.log`: File containing information about R, the OS and attached or loaded packages.
* `bwa/mergedLibrary/macs/<PEAK_TYPE>/consensus/<ANTIBODY>/<COMPARISON>/`
    * `*.results.txt`: Spreadsheet containing comparison-specific DESeq2 output for differential binding results across all peaks.
    * `*FDR0.01.results.txt`, `*FDR0.05.results.txt`: Subset of above file for peaks that pass FDR <= 0.01 and FDR <= 0.05.
    * `*FDR0.01.results.bed`, `*FDR0.05.results.bed`: BED files for peaks that pass FDR <= 0.01 and FDR <= 0.05.
    * `*deseq2.plots.pdf`: MA, Volcano, clustering and scatterplots at FDR <= 0.01 and FDR <= 0.05.
* `bwa/mergedLibrary/macs/<PEAK_TYPE>/consensus/<ANTIBODY>/sizeFactors/`
    * `*.txt`, `*.RData`: Files containing DESeq2 sizeFactors per sample.

</details>


In order to perform the differential binding analysis we need to be able to carry out the read quantification for the same intervals across **all** of the samples in the experiment. To this end, the individual peak-sets called per sample have to be merged together in order to create a consensus set of peaks.

Using the consensus peaks it is possible to assess the degree of overlap between the peaks from a set of samples e.g. *[Which consensus peaks contain peaks that are common/unique to a given set of samples?](https://www.bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/pipeline.html#determine-if-there-is-a-significant-overlap-among-multiple-sets-of-peaks)*. This may be useful for downstream filtering of peaks based on whether they are called in multiple replicates/conditions. Please note that it is possible for a consensus peak to contain multiple peaks from the same sample. Unfortunately, this is sample-dependent but the files generated by the pipeline do have columns that report such instances and allow you to factor them into any further analysis.

![R - Peak intersection plot](`r f<-dir("DiffBind", "vennDiagram.*.png", full.names=TRUE);ifelse(length(f), f[1], "images/r_upsetr_intersect_plot.png")`)

By default, the peak-sets are not filtered, therefore, the consensus peaks will be generated across the union set of peaks from all samples. In future iterations of the pipeline more formal analyses such as [IDR](https://projecteuclid.org/euclid.aoas/1318514284) may be implemented to obtain reproducible and high confidence peak-sets with which to perform this sort of analysis.

The [featureCounts](http://bioinf.wehi.edu.au/featureCounts/) tool is used to count the number of reads relative to the consensus peak-set across all of the samples. This essentially generates a file containing a matrix where the rows represent the consensus intervals, the columns represent all of the samples in the experiment, and the values represent the raw read counts.

![MultiQC - featureCounts consensus peak read assignment plot](`r f<-dir("multiqc_plots/png", "mqc_featureCounts_assignment_plot", full.names=TRUE);ifelse(length(f), f[1], "images/mqc_featureCounts_assignment_plot.png")`)

[DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) is more commonly used to perform differential expression analysis for RNA-seq datasets. However, it can also be used for ChIP-seq differential binding analysis, in which case you can imagine that instead of counts per gene for RNA-seq data we now have counts per bound region.

This pipeline uses a standardized DESeq2 analysis script to get an idea of the reproducibility within the experiment, and to assess the overall differential binding. Please note that this will not suit every experimental design, and if there are other problems with the experiment then it may not work as well as expected. For larger experiments, it may be recommended to use the `vst` transformation instead of the default `rlog` option. See [DESeq2 docs](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization) for a more detailed explanation.


By default, all possible pairwise comparisons across the groups from a particular antibody (as defined in [`design.csv`](#Design-table)) are performed. The DESeq2 results are generated by the pipeline in various ways. You can load up the results across all of the comparisons in a single spreadsheet, or individual folders will also be created that contain the results specific to a particular comparison. For the latter, additional files will also be generated where the intervals have been pre-filtered based on a couple of standard FDR thresholds. Please see [DESeq2 output](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis) for a description of the columns generated by DESeq2.


#### Methods 2: differential by DiffBind

<details markdown="1">
    <summary>Output files</summary>

* `bwa/mergedLibrary/DiffBind/`
    * `DiffBind.res.*.all.csv`: Spreadsheet containing differential binding results across all consensus peaks and all comparisons.
    * `DiffBind.res.*.FDR.05.xls`: Spreadsheet containing differential binding results filtered by FDR < 0.05.
    * `DiffBind.res.*.FDR.05.bedGraph`: BedGraph containing differential binding results filtered by FDR < 0.05.
    * `*.plot.pdf`: Files containing PCA, hierarchical clustering, MA, Volcano plots.
    * `DiffBind.sample.correlation.pdf`: Heatmap plots of sample correlation.
    * `genomicElementDistribuiton*.pdf`: Bar plots of genomic element distribution.
    * `metagenePlotToTSS*.pdf`: Metagene plots of signal around TSS.
    * `sample.csv`: Sample file used in comparison.
    * `chip.rds`: File containing dba.count results.

</details>

In this pipeline, [ChIPpeakAnno](https://bioconductor.org/packages/ChIPpeakAnno/) is used for
peak annotation and quality control. And in downstream, DiffBind is used. [DiffBind](https://bioconductor.org/packages/DiffBind/) is a Bioconductor package to do differential binding analysis of ChIP-seq peak data.

![MultiQC - DiffBind PCA plot](`r ifelse(file.exists("DiffBind/DiffBind.PCA.plot.png"), "DiffBind/DiffBind.PCA.plot.png", "images/mqc_deseq2_pca_plot.png")`)

![MultiQC - DiffBind sample similarity plot](`r ifelse(file.exists("DiffBind/DiffBind.sample.correlation.png"), "DiffBind/DiffBind.sample.correlation.png", "images/mqc_deseq2_sample_similarity_plot.png")`)

![MultiQC - Genomic Element Distribuiton of DiffBind peaks](`r ifelse(file.exists("DiffBind/genomicElementDistribuitonOfDiffBind.png"), "DiffBind/genomicElementDistribuitonOfDiffBind.png", "images/mqc_deseq2_sample_similarity_plot.png")`)

![MultiQC - genomic Element Distribuiton Of all input peaks](`r ifelse(file.exists("DiffBind/genomicElementDistribuitonOfEachPeakList.png"), "DiffBind/genomicElementDistribuitonOfEachPeakList.png", "images/mqc_deseq2_sample_similarity_plot.png")`)

![MultiQC - Diffbind volcano plot](`r f<-dir("DiffBind", "*Volcano.plot.png", full.names=TRUE);ifelse(length(f), f[1], "images/r_deseq2_volcano_plot.png")`)

## Pipeline information

<details markdown="1">
    <summary>Output files</summary>

* `pipeline_info/`
    * `pipeline_report.html`, `pipeline_report.txt`, `software_versions.csv`: Reports generated by the pipeline.
    * `execution_report.html`, `execution_timeline.html`, `execution_trace.txt`, `pipeline_dag.svg`: Reports generated by Nextflow.
    * `design_reads.csv`, `design_controls.csv`: Reformatted design files used as input to the pipeline.
    * `results_description.html`: Documentation for interpretation of results in HTML format.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to trouble-shoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

### Package information

```{r, echo=FALSE}
pkgs <- installed.packages()
datatable(pkgs)
```

### software_versions

```{r, results = 'asis', echo=FALSE}
files <- dir("software_versions", "yaml", full.names = TRUE)
for(i in files){
  data <- yaml::yaml.load_file(i)
  data <- strsplit(data$data, "\n")[[1]]
  data <- data[gsub("\\s+", "", data)!=""]
  cat(data, sep = "\n")
}
```


### MD5 checksum

MD5 checksum is required for [Gene Expression Omnibus (GEO)](https://www.ncbi.nlm.nih.gov/geo/info/submission.html) submission. 
The checksums listed here is the checksum for un-compresed fastq files.

```{r, echo=FALSE}
f <- dir(".", "md5.*.txt")
d <- unlist(lapply(f, readLines, warn = FALSE))
data <- do.call(rbind, strsplit(d, "\\s+"))
datatable(data)
```



